elastic/kibana
# [RAC][Alerts] - Addition of RBAC to unified alerts index

## Summary

++ in collaboration with @dhurley14 

This PR adds RBAC (Role based access-control) for the `.alerts` index that rules from numerous features/plugins will be writing into. 

Initial issue outlining work [95721](https://github.com/elastic/kibana/issues/95721).

The below diagram is a very rough outline of what we plan to do, based largely off of RBAC work being done by Cases and existing RBAC code done by Alerting.

<img width="1341" alt="alerts as data role based access control authorization request flow diagram" src="https://user-images.githubusercontent.com/915763/114610203-a2b3e900-9c6d-11eb-819e-b8d676d42041.png">

The alerts RBAC will be based off of the alerting plugin's existing RBAC model. We will be using the same security plugin [`Actions`](https://github.com/elastic/kibana/blob/master/x-pack/plugins/security/server/authorization/actions/alerting.ts) and [`FeaturePrivilegeBuilder`](https://github.com/elastic/kibana/blob/master/x-pack/plugins/security/server/authorization/privileges/feature_privilege_builder/alerting.ts). What does all that mean? Actions in the security plugin are used to build out URIs that are then associated to users to determine what they have access to. So for alerting, a user who has access to get a rule from the security solution may have a URI like `alerting:DetectionEngine/securitySolution/get`. What we want to do is reuse this, but add an additional piece of information in there to denote whether the action is referring to a rule or an alert. The following URIs associated with alerts would be:

##### Get an alert
`alerting:<rule-alert-type>/<producer>/alert/get`

##### Search alerts
`alerting:<rule-alert-type>/<producer>/alert/find`

##### Update an alert
`alerting:<rule-alert-type>/<producer>/alert/updateStatus`

In addition to reusing the `Actions` and `FeaturePrivilegeBuilder`, the RAC auth class will extend the [alerting auth class](https://github.com/yctercero/kibana/blob/master/x-pack/plugins/alerting/server/authorization/alerts_authorization.ts) to reuse the following methods: `ensureAuthorized` and `getFindAuthorizationFilter`. The `ensureAuthorized` is reused nearly 1 for 1, however the `getFindAuthorizationFilter` is modified slightly to allow us to ask for an ES filter, as opposed to a kuery built around the SO structure. This is because the alerts are ES entities and rules are SO/Kibana entities. Each is searched differently, so the filter and what is appended to it varies slightly.

### Kibana Feature Privilege
Within the feature privilege object, these privileges would be displayed as follows (note that this may differ depending on how open we want to leave it to users to have varying combinations of read/write between rules and alerts):
```js
plugins.features.registerKibanaFeature({
      id: MY_APP_ID,
      name: 'My App',
      alerting:['siem.signals'],
      privileges: {
        all: {
          alerting: {
            all:['siem.signals'],
          },
        },
        read: {
          alerting: {
            read:['siem.signals'],
          },
        },
      },
    subFeatures: [
      {
        name:  'My App',
        privilegeGroups: [
          {
            groupType: 'independent',
            privileges: [
              {
                id: 'allowAlertStatusUpdate',
                name: 'Alert status update',
                includeIn: 'all',
                alerting: {
                  all: ['alertUpdate'],
                },
              },
            ],
          },
        ],
      },
    ],
});
```
This means that a user who has a Kibana role with Security Solution set to `All` would get `all` privileges for rules and alerts, a role set to `Read` would get `read` privileges for rules and alerts, and if they would like, they can select to modify the subfeature to allow a user to be `read` on rules and `write` on alerts.

![image](https://user-images.githubusercontent.com/10927944/116760750-a4e6a900-a9ca-11eb-94fa-8c9ede6921d1.png)

## Testing stuff out

Pull down [this branch](https://github.com/elastic/elasticsearch/pull/72181) and run ES locally using the following steps..

Running ES locally
1. Edit `buildSrc/src/main/groovy/elasticsearch.run.gradle` and add setting `'xpack.security.authc.api_key.enabled', 'true'` after line 24
2. `./gradlew run` # this runs with a trial license
3. Execute below curl script to post `kibana_elastic` user
```bash
curl -u elastic:password -X POST "http://127.0.01:9200/_security/user/kibana_elastic?pretty" -H 'Content-Type: application/json' -d '{"password":"changeme","roles":["superuser"],"full_name":"kibana","email":"jacknich@example.com"}'
```
4. Set `kibana.dev.yml` to use `kibana_elastic` as the user
```yaml
elasticsearch:
  username: 'kibana_elastic'
  password: 'changeme'
  hosts: 'http://127.0.0.1:9200'
```
Start up kibana


You should now be able to create a rule which generates alerts and then "find" those alerts (using the kibana system user) using the scripts located in `x-pack/plugins/rule_registry/server/scripts/get_security_solution_alert.sh`


<details><summary>Test data</summary>

```
POST myfakeindex-1/_doc
{
  "message": "hello world 1"
}

POST myfakeindex-2/_doc
{
  "message": "hello world 2",
  "event": {
    "ingested": "2021-04-30T15:23:03.520Z" <replace with current ISO date>
  }
}

POST myfakeindex-3/_doc
{
  "message": "hello world 3",
  "@timestamp": "2021-04-30T15:23:03.520Z" <replace with current ISO date>
}
```
</details>

Create a rule to query `myfa*` and it should generate an alert in the security solution

To get the alert, change directory into `x-pack/plugins/rule_registry/server/scripts` and execute

```bash
term$ ./get_security_solution_alert.sh
```

This script will post the security role and observer role and execute a find using the new alerts as data client. To test the authz functionality execute the below, expecting a 403 response.

```bash
term$ ./get_security_solution_alert.sh observer
term$ {
  "statusCode": 401,
  "error": "Unauthorized",
  "message": "Unauthorized to get \"rac:8.0.0:siem/get\" alert\""
}
```

### Checklist
